### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by afk11**
Work in progress for now, feedback on where I'm moving things welcome!

### CODE REVIEW

File: src/Math/Gmp.php [hunk 1]
```diff
@@ -226,6 +227,8 @@ public function intToString($x)
 
             return $result;
         }
+
+        throw new \InvalidArgumentException('Unable to convert int to string');
```
Comment (btcdrak) [inline src/Math/Gmp.php]: Looks more like a LogicException/RuntimeException then.

File: src/Crypto/Key/KeyPair.php [hunk 2]
```diff
@@ -31,12 +34,32 @@ public function __construct(PrivateKeyInterface $privateKey = null, PublicKeyInt
         $this->privateKey = $privateKey;
         $this->publicKey = $publicKey;
 
-        if ($this->privateKey !== null && $this->publicKey !== null && ! $this->publicKey->equals($privateKey->getPublicKey())) {
+        // TODO! PublicKey::equals() does not exist
+        if ($this->privateKey !== null && $this->publicKey !== null
+        //    && ! $this->publicKey->equals($privateKey->getPublicKey())
```
Comment (afk11) [inline src/Crypto/Key/KeyPair.php]: Looks like PublicKey::equals() was removed, or never existed. Probably a good idea to add it..

File: .gitignore [hunk 3]
```diff
@@ -1,3 +1,4 @@
+.idea
```
Comment (btcdrak) [inline .gitignore]: This belongs in your local global git ignore https://help.github.com/articles/ignoring-files/#create-a-global-gitignore otherwise the repo .gitignore needs to contain all the cases for everyone's setup. If you have a local global gitignore all your projects are handled with your own personal IDE preferences.

File: src/Crypto/Key/KeyPair.php (hunk 2 above)
Comment (btcdrak) [inline src/Crypto/Key/KeyPair.php]: Yes.

**Comment by btcdrak** [comment]
Apart from nits, PR looks good to me. Any one else got any comments?

**Comment by aztech-dev** [comment]
ACK for me. Like the decluttering of root NS....

File: src/Crypto/Key/KeyPair.php (hunk 2 above)
Comment (aztech-dev) [inline src/Crypto/Key/KeyPair.php]: Agreed. It's not critical as I don't think KeyPair is currently used, but always a nice addition/fix.

**Comment by afk11** [comment]
Cheers, I still have one or two more bits to do. I merged @aztech-dev 's forked branch on rfc6979 tests - looks like one test case fails but with the amount of them I guess it could be a copy/paste issue.

Aye root NS is quite empty right now, which is good, hopefully the new home for everything makes sense? 

I have been pondering some more changes also, what do people think of doing gmp_init() on their input? In terms of the speed thing, doing gmp_strval after every single math operation (I PR'd that ages ago for consistency...) probably drags it down a lot. 

An small improvement _might_ be that every function gmp_init()'s its input, and does gmp_strval on the output. So then we go from tens of gmp_strvals to only a small number per function. Problem is that lots of public facing functions use others. So, why don't we just do gmp_init on all our inputs, and assume that resources will be returned everywhere, so it's only done at the very start of your routine, and strval at the very end? 

On a similar note, we could just wrap these values as a GmpInt() - something @rubensayshi suggested.

Also one last thing to explore would be a class without the conditional swap for some small niche cases. This probably needs careful planning. Signer::verify() is a good place for this, as anyone can do that without a secret. Signer::sign() does (nonce*G) which probably does require a constant time algo since the nonce must be unknown. 

Will get back to this later tonight or tomorrow, would appreciate any feedback or tangential ideas to the above :)

**Comment by aztech-dev** [comment]
Awesome for merging my changes -- haven't had too much time to dwelve back into the branch, though I definitely wanted to get it done.

I'm not too fond of placing the burden of conversion on the library user, I'd rather see a GmpInt class. It's definitely a better approach (the Math classes were always in my eyes a means to transition from the original codebase to a better approach -- I originally stuffed in there everything that called either bc or gmp to isolate the calls and removing all the code duplication that I could find). It would also be smart to see how PHP7 handles the code with/without gmp_strval's. If I recall correctly, PHP 7 is supposed to handle bigint's transparently now (any int greater than 32bits is internally (? not sure here) cast to a gmp resource).

I'm not however convinced the performance boost will be tremendous -- the most expensive operation is obviously point multiplication, and the current algo relies on base conversions -- which boil down to gmp_init/gmp_strval -- on every iteration. But any venue for a significant perf improvement is worth exploring, and I'd love to be proved wrong (or see a better performing implementation of the montgomery ladder).

Also regarding the constant time algos, sounds like a good idea to provide a constant and a non-constant time implementation. Just need to make sure that it's done so that library consumers can't make a mistake easily.

**Comment by afk11** [comment]
Yeah I recall - if()'s in pretty much every function. Definitely a
worthwhile change.

I think Ruben mentioned there was a 30% increase removing the
gmp_strvals, on some function, but it wasn't a large refactoring, so not
a full picture of how it might go.

Hmm, actually while I think of it, I was talking to someone recently who
mentioned that double-and-add is normally the slowest algo; and that
others are generally quicker. If base conversion is indeed the
bottleneck, perhaps we'll see a performance GAIN by using the Montgomery
ladder, and not be forced into using a non-constant time multiplication
occasionally? Will rethink things before I plough ahead anyway..

PHP7 will be exciting, it'll be good to see how a native implementation
compares I guess.

File: .gitignore (hunk 3 above)
Comment (afk11) [inline .gitignore]: Nice, thanks for that!

File: composer.json [hunk 4]
```diff
@@ -29,12 +29,9 @@
         "ext-gmp": "*",
         "ext-mcrypt": "*",
         "fgrosse/phpasn1": "~1.1.0",
-        "symfony/console": "~2.6"
-    },
-    "require-dev": {
```
Comment (afk11) [inline composer.json]: Removed require-dev to silence composer errors

File: composer.json [hunk 5]
```diff
@@ -29,12 +29,9 @@
...
+        "symfony/console": "~2.6",
         "phpunit/phpunit": "~4.1",
         "squizlabs/php_codesniffer": "~2",
-        "padraic/phpunit-extensions": "~1.0@dev",
```
Comment (afk11) [inline composer.json]: dev stability, doesn't seem to be used for anything..

**Comment by afk11** [comment]
Notable changes, aside from mass moving around:
PublicKey/PrivateKey classes moved to Mdanter\Ecc\Crypto\Key
Signature namespace mvoed to Mdanter\Ecc\Crypto\Routines\Signature
EcDH has also been moved to the Routines namespace
PrivateKey::getPoint() renamed to PrivateKey::getGenerator()

Signer now has truncateHash(GeneratorPoint $G, $messageHash) now truncates/pads hashes when being used for ECDSA. 

I exposed PrimeFieldArithmetic, NumberTheory, and EcMath in the Gmp class also, and added getSigner() to EccFactory. I can't spot much more that might be useful there, so ideas welcome.

@aztech-dev I am thinking the signer class could make use of PrimeFieldArithmetic, though it seems to use the curves order instead of the prime. Might rename this to ModularArithmetic? Or what do you think here? 

Also asking because EcMath (operates on points or integers) currently requires you do mod() occasionally - and thinking it could just use ModularArithmetic and safely operate with integers/points without requiring the user to use EcMaths mod()? They seem to fit in nicely together.

**Comment by btcdrak** [comment]
+1

File: src/Crypto/EcDH/EcDH.php [hunk 6]
```diff
@@ -117,11 +124,11 @@ public function setSenderKey(PrivateKeyInterface $key)
      * {@inheritDoc}
      * @see \Mdanter\Ecc\EcDHInterface::encrypt()
      */
-    public function encrypt($string)
+    public function encrypt(Message $string)
```
Comment (aztech-dev) [inline src/Crypto/EcDH/EcDH.php]: Maybe rename $string into $message ? Just nitpicking :p

File: src/Crypto/EcDH/EcDHInterface.php [hunk 7]
```diff
@@ -64,16 +66,16 @@ public function setRecipientKey(PublicKeyInterface $key);
      * Generates an encrypted version of the given text, that can be decoded by the other
      * party.
      *
-     * @param  string $string
+     * @param Message $string $string
      * @return string
      */
-    public function encrypt($string);
+    public function encrypt(Message $string);
```
Comment (aztech-dev) [inline src/Crypto/EcDH/EcDHInterface.php]: Same rename here...

**Comment by aztech-dev** [comment]
Looks good overall, my only issue is with EccFactory::getMessage. I'm generally against static calls inside a code base, as it introduces strong coupling, and makes it impossible to mock out the static calls. 

A better alternative would probably be to have MessageFactory which can be instantiated and injected into the classes that depend upon it, though I don't consider it a blocker for now.

I'm also starting to think the curves should convey info about the key size instead of having to infer it from the actual size of the key (ref. point multiplication and signing). I may be mistaken, but I think our current strategy may trigger some non-constant time calculations in places where constant time is required, but again, that's not a blocker for now.

So ACK from me (minus those arg names)

File: .gitignore (hunk 3 above)
Comment (btcdrak) [inline .gitignore]: Did this get missed?

**Comment by aztech-dev** [comment]
PS. As far as I'm concerned, KeyPair is not used and not necessary (dev artefact from me)

File: src/Crypto/EcDH/EcDH.php (hunk 6 above)
Comment (afk11) [inline src/Crypto/EcDH/EcDH.php]: Fine by me :D

**Comment by afk11** [comment]
Ok, will go over it with that in mind! Main effort this week was just getting tests to pass, so will mop up the last of those today. I could probably drop that getMessage function, instantiating it is just as easy.

I'd agree the key size should be known to the curve, I was wondering a bit about why the EcDH shared secret has hashed, it's to avoid the high bits being zero. Might start following some of this up as separate PR's :) 

I wonder should we introduce a class for the encrypted data? Maybe MessageFactory can produce an EncryptedMessage? 

Just BTW, I passed sha256 as the algorithm for messages in EcDH, I'm not entirely sure how correct this is, but figured since we're hashing the key with that.. Ideas?

**Comment by afk11** [comment]
Noticed a bunch of the doc comments are screwed after refactoring. Will give it a once over later to fix that. What else is on the cards for 0.3.0?

**Comment by aztech-dev** [comment]
Not against the EncryptedMessage class, but I think current P/R is big enough :)

I'm with you on passing the hash algo as a param to MessageFactory. sha256 is what the lib originally used in EcDH, but IIRC there's no constraint on the hash algo used (though it's always a good idea to use a hash size similar to the size of the key).

For 0.3.0 in terms of BC breaks I think everything's covered, there are a few more things in the milestone, but not much (doc, extra console features, some extra tests). Only critical thing is the gmp_strval issue, though I'm not sure/convinced it's a good fit for 0.3.0 considering the amount of work involved.

**Comment by btcdrak** [comment]
@afk11 there is a ticket for 0.3.0 release.

**Comment by afk11** [comment]
Damn - seems there's still an edge case in HmacDrbg.. All tests pass bar those for nist-p521, which is the only one not to have a curve where n % 8 == 0. 

There is definitely an inconsistency with the RFC, but for the sake of the feature which is to yield an unpredictable k.. I will revert these test vectors for now, so we can merge and I can get on with the rest. 

The fix is likely just truncating/padding, so it'll be non-breaking change when it comes.

**Comment by btcdrak** [comment]
@afk11 might be an idea to ask on #bitcoin-dev if they are aware of any particular issues with the RFC. @gmaxwell or @sipa might have an idea.

Your PR will need rebased also.

**Comment by afk11** [comment]
I will, though this failure is almost surely a fault in this implementation anyways, not the RFC, as there are official test vectors for the curve we're having trouble with. Guess there's a shortcut somewhere that isn't right..

**Comment by afk11** [comment]
Rebasing was messy - but all good to merge this now. Will tackle other issues then before we tag it!

**Comment by aztech-dev** [comment]
Just catching up on all the changes... you removed the P521 test vectors because they weren't correct ? Or another issue ?

**Comment by afk11** [comment]
Removed them because they didn't pass, and on my todo list. Guessing it's a bit count issue or something, but the fix should be non-breaking.

**Comment by aztech-dev** [comment]
Hum ok. Not too fond of the idea of removing tests because they don't pass (as a library consumer, if I see green tests, I'll wrongly assume everything is fine when in fact it is not), so don't forget to put them back at some point :)

**Comment by btcdrak** [comment]
Maybe we should open a PR with the failed tests so we know there's something pending?

**Comment by aztech-dev** [comment]
Allright, I'll issue one tonight unless @afk11 beats me to it
